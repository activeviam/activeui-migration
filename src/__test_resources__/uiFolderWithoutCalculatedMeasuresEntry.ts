/**
 * An already migrated UI folder that is missing an entry for `calculated_measures`.
 * Useful for unit tests.
 */
export const uiFolderWithoutCalculatedMeasuresEntry = {
  entry: {
    isDirectory: true,
    owners: ["ROLE_CS_ROOT"],
    readers: ["ROLE_CS_ROOT"],
  },
  children: {
    filters: {
      entry: {
        isDirectory: true,
        owners: ["ROLE_CS_ROOT"],
        readers: ["ROLE_CS_ROOT"],
      },
      children: {
        content: {
          children: {},
          entry: {
            isDirectory: true,
            owners: ["ROLE_USER"],
            readers: ["ROLE_USER"],
          },
        },
        structure: {
          children: {},
          entry: {
            isDirectory: true,
            owners: ["ROLE_USER"],
            readers: ["ROLE_USER"],
          },
        },
      },
    },
    widgets: {
      entry: {
        isDirectory: true,
        owners: ["ROLE_CS_ROOT"],
        readers: ["ROLE_CS_ROOT"],
      },
      children: {
        content: {
          children: {},
          entry: {
            isDirectory: true,
            owners: ["ROLE_USER"],
            readers: ["ROLE_USER"],
          },
        },
        structure: {
          children: {},
          entry: {
            isDirectory: true,
            owners: ["ROLE_USER"],
            readers: ["ROLE_USER"],
          },
        },
      },
    },
    dashboards: {
      entry: {
        isDirectory: true,
        owners: ["ROLE_CS_ROOT"],
        readers: ["ROLE_CS_ROOT"],
      },
      children: {
        content: {
          children: {
            bef: {
              entry: {
                content:
                  '{"pages":{"p-0":{"layout":{"children":[{"leafKey":"0","size":1}],"direction":"row"},"name":"Page 1","filters":[],"content":{"0":{"query":{"mdx":"SELECT FROM [EquityDerivativesCube] CELL PROPERTIES VALUE, FORMATTED_VALUE, BACK_COLOR, FORE_COLOR, FONT_FLAGS","updateMode":"once"},"filters":[],"queryContext":[],"mapping":{"rows":[],"columns":["ALL_MEASURES"],"measures":[]},"name":"Pivot table with calculated measures","serverKey":"my-server","widgetKey":"tree-table","columnWidths":{}}},"queryContext":[]}},"pagesOrder":["p-0"],"filters":[],"queryContext":[]}',
                isDirectory: false,
                owners: ["admin"],
                readers: ["admin"],
                timestamp: 1658155953186,
                lastEditor: "admin",
                canRead: true,
                canWrite: true,
              },
            },
          },
          entry: {
            isDirectory: true,
            owners: ["ROLE_USER"],
            readers: ["ROLE_USER"],
          },
        },
        structure: {
          children: {
            bef: {
              entry: {
                isDirectory: true,
                owners: ["admin"],
                readers: ["admin"],
                timestamp: 1658155953644,
                lastEditor: "admin",
                canRead: true,
                canWrite: true,
              },
              children: {
                bef_metadata: {
                  entry: {
                    isDirectory: true,
                    owners: ["admin"],
                    readers: ["admin"],
                    timestamp: 1658155953644,
                    lastEditor: "admin",
                    canRead: true,
                    canWrite: true,
                    content: '{"name":"Calculated measures dashboard."}',
                  },
                },
              },
            },
          },
          entry: {
            isDirectory: true,
            owners: ["ROLE_USER"],
            readers: ["ROLE_USER"],
          },
        },
        thumbnails: {
          children: {},
          entry: {
            isDirectory: true,
            owners: ["ROLE_USER"],
            readers: ["ROLE_USER"],
          },
        },
      },
    },
    organization_settings: {
      entry: {
        content: "{}",
        owners: ["ROLE_CS_ROOT"],
        readers: ["ROLE_USER"],
      },
    },
    organization_permissions: {
      entry: {
        isDirectory: false,
        owners: ["ROLE_CS_ROOT"],
        readers: ["ROLE_CS_ROOT"],
        content: "{}",
      },
    },
    user_roles: {
      entry: {
        isDirectory: true,
        owners: ["ROLE_CS_ROOT"],
        readers: ["ROLE_USER"],
      },
    },
    users: {
      entry: {
        isDirectory: true,
        owners: ["ROLE_USER"],
        readers: ["ROLE_USER"],
      },
      children: {
        admin: {
          entry: {
            isDirectory: true,
            owners: ["admin"],
            readers: ["admin"],
            timestamp: 1658155953761,
            lastEditor: "admin",
            canRead: true,
            canWrite: true,
          },
          children: {
            settings: {
              entry: {
                content: "{}",
                owners: ["admin"],
                readers: ["admin"],
              },
            },
            activity: {
              entry: {
                content: "{}",
                owners: ["admin"],
                readers: ["admin"],
              },
            },
          },
        },
      },
    },
  },
};
