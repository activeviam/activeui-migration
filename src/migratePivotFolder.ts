import { ContentRecord } from "@activeviam/activeui-sdk";
import xml2js from "xml2js";
import _cloneDeep from "lodash/cloneDeep";
import { generateId } from "./generateId";

/**
 * Extracts and parses the XML calculated measure objects from the given pivot folder.
 */
const getCalculatedMeasuresFromPivotFolder = async (
  pivotFolder: ContentRecord
): Promise<
  {
    expression: string;
    formatStringExpression: string;
    uniqueName: string;
    id: string;
  }[]
> => {
  const parser = new xml2js.Parser();

  const calculatedMeasures: {
    expression: string;
    formatStringExpression: string;
    uniqueName: string;
    id: string;
  }[] = [];

  if (pivotFolder?.children) {
    await Promise.all(
      Object.values(pivotFolder?.children).map(async ({ children = {} }) => {
        const calculatedMeasureEntries = Object.entries(children);

        const measures = await Promise.all(
          calculatedMeasureEntries.map(async (cmEntry) => {
            const [_, entry] = cmEntry;
            const result = await parser.parseStringPromise(entry.entry.content);
            const id = generateId();

            return { id, ...result.calculatedMember["$"] };
          })
        );

        calculatedMeasures.push(...measures);
      })
    );
  }

  return calculatedMeasures;
};

/**
 * Returns the migrated UI folder, along with the converted saved calculated measures from their location in ActiveUI 4 to their respective location in ActiveUI 5.
 */
export async function migratePivotFolder(
  legacyPivotFolder: ContentRecord,
  migratedUIFolder: ContentRecord
) {
  const entitlements = legacyPivotFolder.children?.entitlements;
  const calculatedMeasuresRoot = entitlements?.children?.cm;
  const migratedUIFolderWithCalculatedMeasures = _cloneDeep(migratedUIFolder);

  const calculatedMeasuresToMigrate = calculatedMeasuresRoot
    ? await getCalculatedMeasuresFromPivotFolder(calculatedMeasuresRoot)
    : undefined;

  const migratedCalculatedMeasuresEntry =
    (migratedUIFolderWithCalculatedMeasures.children &&
      migratedUIFolderWithCalculatedMeasures.children.calculated_measures) || {
      entry: { isDirectory: true },
      children: {
        content: { children: {} },
        structure: { children: {} },
      },
    };

  // The entry for the calculated measures already exists, just add them under children.content and children.structure

  const calculatedMeasuresNames = calculatedMeasuresToMigrate?.map(
    (measure) => ({ [measure.id]: measure.uniqueName })
  );
  const measuresContentEntries = calculatedMeasuresToMigrate
    ? calculatedMeasuresToMigrate
        .map((measure) => {
          return {
            [measure.id]: {
              entry: {
                content: JSON.stringify({
                  expression: measure.expression,
                  properties: [
                    `FORMAT_STRING=${measure.formatStringExpression}`,
                  ],
                }),
                isDirectory: false,
              },
            },
          };
        })
        .reduce((acc, currentCM) => {
          return { ...acc, ...currentCM };
        }, {})
    : {};

  const measuresStructureEntries = (calculatedMeasuresNames || [])
    ?.map((measure) => {
      const [[id, name]] = Object.entries(measure);

      return {
        [id]: {
          entry: {
            isDirectory: true,
          },
          children: {
            [`${id}_metadata`]: {
              entry: {
                content: JSON.stringify({
                  name,
                }),
                isDirectory: false,
              },
            },
          },
        },
      };
    })
    .reduce((acc, currentCM) => {
      return { ...acc, ...currentCM };
    }, {});

  const migratedMeasuresContent =
    // @ts-ignore This is safe to ingore due to the typecheck above that ensures that `migratedCalculatedMeasuresEntry.children` will not be undefined.
    migratedCalculatedMeasuresEntry.children.content;

  const migratedMeasuresStructure =
    // @ts-ignore This is safe to ingore due to the typecheck above that ensures that `migratedCalculatedMeasuresEntry.children` will not be undefined.
    migratedCalculatedMeasuresEntry.children.structure;

  migratedMeasuresStructure.children = {
    ...migratedMeasuresStructure.children,
    ...measuresStructureEntries,
  };

  migratedMeasuresContent.children = {
    ...migratedMeasuresContent.children,
    ...measuresContentEntries,
  };

  migratedUIFolderWithCalculatedMeasures.children = {
    ...migratedUIFolderWithCalculatedMeasures.children,
    calculated_measures: {
      entry: {
        isDirectory: true,
        readers: ["ROLE_USER"],
        owners: ["ROLE_USER"],
        ...(migratedUIFolderWithCalculatedMeasures.children?.calculated_measures
          ?.entry || {}),
      },
      children: {
        content: {
          entry: {
            isDirectory: true,
            readers: ["ROLE_USER"],
            owners: ["ROLE_USER"],
          },
          ...migratedMeasuresContent,
        },
        structure: {
          entry: {
            isDirectory: true,
            readers: ["ROLE_USER"],
            owners: ["ROLE_USER"],
          },
          ...migratedMeasuresStructure,
        },
      },
    },
  };

  return migratedUIFolderWithCalculatedMeasures;
}
