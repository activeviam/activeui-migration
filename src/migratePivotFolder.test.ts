import { ContentRecord } from "@activeviam/activeui-sdk";
import { migratePivotFolder } from "./migratePivotFolder";
import { migrateUIFolder } from "./migrateUIFolder";
import { servers } from "./__test_resources__/servers";
import { smallLegacyPivotFolder } from "./__test_resources__/smallLegacyPivotFolder";
import { smallLegacyUIFolder } from "./__test_resources__/smallLegacyUIFolder";
import { uiFolderWithoutCalculatedMeasuresEntry } from "./__test_resources__/uiFolderWithoutCalculatedMeasuresEntry";

describe("migratePivotFolder", () => {
  beforeAll(() => {
    const generateId = require("./generateId");
    let counter = 0;
    jest.spyOn(generateId, "generateId").mockImplementation(() => {
      const id = `00${counter}`;
      counter += 1;
      return id;
    });
  });

  it("returns a valid ActiveUI5 /ui folder", async () => {
    const migratedUIFolder = migrateUIFolder(smallLegacyUIFolder, servers);
    const migratedUIFolderWithCalculatedMeasures = await migratePivotFolder(
      smallLegacyPivotFolder,
      migratedUIFolder
    );

    expect(migratedUIFolderWithCalculatedMeasures).toMatchSnapshot();
  });

  it("returns a valid ActiveUI5 /ui folder on an input that does not contain an entry for calculated measures", async () => {
    const migratedUIFolderWithCalculatedMeasures = await migratePivotFolder(
      smallLegacyPivotFolder,
      uiFolderWithoutCalculatedMeasuresEntry
    );

    expect(migratedUIFolderWithCalculatedMeasures).toMatchSnapshot();
  });
});
