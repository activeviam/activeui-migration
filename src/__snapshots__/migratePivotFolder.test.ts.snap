// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`migratePivotFolder returns a valid ActiveUI5 /ui folder 1`] = `
Object {
  "children": Object {
    "calculated_measures": Object {
      "children": Object {
        "content": Object {
          "children": Object {
            "000": Object {
              "entry": Object {
                "content": "{\\"expression\\":\\"IIf(IsEmpty(([Booking].[Desk].CurrentMember.Parent, [Measures].[pnl.SUM])), null, [Measures].[pnl.SUM] / ([Booking].[Desk].CurrentMember.Parent, [Measures].[pnl.SUM]))\\",\\"properties\\":[\\"FORMAT_STRING=\\\\\\"#,###.##%\\\\\\"\\"]}",
                "isDirectory": false,
              },
            },
            "001": Object {
              "entry": Object {
                "content": "{\\"expression\\":\\"IIf(IsEmpty(([Currency].[Currency].[ALL].[AllMember], [Measures].[pnl.FOREX])), null, [Measures].[pnl.FOREX] / ([Currency].[Currency].[ALL].[AllMember], [Measures].[pnl.FOREX]))\\",\\"properties\\":[\\"FORMAT_STRING=\\\\\\"#,###.##%\\\\\\"\\"]}",
                "isDirectory": false,
              },
            },
            "002": Object {
              "entry": Object {
                "content": "{\\"expression\\":\\"Count(Descendants([CounterParty].[CounterParty].CurrentMember, [CounterParty].[CounterParty].[CounterPartyGroup]), EXCLUDEEMPTY)\\",\\"properties\\":[\\"FORMAT_STRING=\\\\\\"#,###.##\\\\\\"\\"]}",
                "isDirectory": false,
              },
            },
          },
          "entry": Object {
            "isDirectory": true,
            "owners": Array [
              "ROLE_USER",
            ],
            "readers": Array [
              "ROLE_USER",
            ],
          },
        },
        "structure": Object {
          "children": Object {
            "000": Object {
              "children": Object {
                "000_metadata": Object {
                  "entry": Object {
                    "content": "{\\"name\\":\\"[Measures].[third calculated measure]\\"}",
                    "isDirectory": false,
                  },
                },
              },
              "entry": Object {
                "isDirectory": true,
              },
            },
            "001": Object {
              "children": Object {
                "001_metadata": Object {
                  "entry": Object {
                    "content": "{\\"name\\":\\"[Measures].[second calculated measure]\\"}",
                    "isDirectory": false,
                  },
                },
              },
              "entry": Object {
                "isDirectory": true,
              },
            },
            "002": Object {
              "children": Object {
                "002_metadata": Object {
                  "entry": Object {
                    "content": "{\\"name\\":\\"[Measures].[first calculated measure]\\"}",
                    "isDirectory": false,
                  },
                },
              },
              "entry": Object {
                "isDirectory": true,
              },
            },
          },
          "entry": Object {
            "isDirectory": true,
            "owners": Array [
              "ROLE_USER",
            ],
            "readers": Array [
              "ROLE_USER",
            ],
          },
        },
      },
      "entry": Object {
        "isDirectory": true,
        "owners": Array [
          "ROLE_CS_ROOT",
        ],
        "readers": Array [
          "ROLE_CS_ROOT",
        ],
      },
    },
    "dashboards": Object {
      "children": Object {
        "content": Object {
          "children": Object {
            "158": Object {
              "entry": Object {
                "canRead": true,
                "canWrite": true,
                "content": "{\\"pages\\":{\\"p-0\\":{\\"layout\\":{\\"children\\":[{\\"leafKey\\":\\"2\\"},{\\"leafKey\\":\\"1\\"}],\\"direction\\":\\"row\\"},\\"name\\":\\"Page 1\\",\\"filters\\":[],\\"content\\":{\\"1\\":{\\"query\\":{\\"mdx\\":\\"SELECT NON EMPTY Crossjoin(Hierarchize(DrilldownLevel([Geography].[City].[ALL].[AllMember])), Hierarchize(DrilldownLevel([Currency].[Currency].[ALL].[AllMember]))) ON ROWS, NON EMPTY [Measures].[contributors.COUNT] ON COLUMNS FROM [EquityDerivativesCube] CELL PROPERTIES VALUE, FORMATTED_VALUE, BACK_COLOR, FORE_COLOR, FONT_FLAGS\\",\\"updateMode\\":\\"once\\"},\\"filters\\":[],\\"queryContext\\":[{\\"key\\":\\"mdx.hiddengrandtotals\\",\\"value\\":\\"1\\"}],\\"mapping\\":{\\"rows\\":[\\"[Geography].[City].[City]\\",\\"[Currency].[Currency].[Currency]\\"],\\"columns\\":[\\"ALL_MEASURES\\"],\\"measures\\":[\\"[Measures].[contributors.COUNT]\\"]},\\"name\\":\\"Untitled Pivot Table\\",\\"serverKey\\":\\"my-server\\",\\"widgetKey\\":\\"tree-table\\",\\"columnWidths\\":{}},\\"2\\":{\\"switchedTo\\":\\"plotly-clustered-column-chart\\",\\"mapping\\":{\\"xAxis\\":[\\"[Currency].[Currency].[Currency]\\"],\\"values\\":[\\"[Measures].[pnl.FOREX]\\"],\\"secondaryValues\\":[],\\"splitBy\\":[\\"[Booking].[Desk].[LegalEntity]\\",\\"ALL_MEASURES\\"],\\"horizontalSubplots\\":[],\\"verticalSubplots\\":[]},\\"query\\":{\\"mdx\\":\\"SELECT NON EMPTY Crossjoin(Hierarchize(DrilldownLevel([Currency].[Currency])), Hierarchize(DrilldownLevel([Booking].[Desk].[ALL].[AllMember]))) ON ROWS, NON EMPTY [Measures].[pnl.FOREX] ON COLUMNS FROM [EquityDerivativesCube]\\",\\"updateMode\\":\\"once\\"},\\"filters\\":[],\\"queryContext\\":[],\\"serverKey\\":\\"my-server\\",\\"name\\":\\"Untitled Chart\\",\\"widgetKey\\":\\"plotly-line-chart\\"}},\\"queryContext\\":[]}},\\"pagesOrder\\":[\\"p-0\\"],\\"filters\\":[],\\"queryContext\\":[]}",
                "isDirectory": false,
                "lastEditor": "admin",
                "owners": Array [
                  "admin",
                ],
                "readers": Array [
                  "admin",
                ],
                "timestamp": 1607879735685,
              },
            },
          },
          "entry": Object {
            "isDirectory": true,
            "owners": Array [
              "ROLE_USER",
            ],
            "readers": Array [
              "ROLE_USER",
            ],
          },
        },
        "structure": Object {
          "children": Object {
            "158": Object {
              "children": Object {
                "158_metadata": Object {
                  "entry": Object {
                    "canRead": true,
                    "canWrite": true,
                    "content": "{\\"name\\":\\"hidden grand totals\\"}",
                    "isDirectory": true,
                    "lastEditor": "admin",
                    "owners": Array [
                      "admin",
                    ],
                    "readers": Array [
                      "admin",
                    ],
                    "timestamp": 1607879735685,
                  },
                },
              },
              "entry": Object {
                "canRead": true,
                "canWrite": true,
                "isDirectory": true,
                "lastEditor": "admin",
                "owners": Array [
                  "admin",
                ],
                "readers": Array [
                  "admin",
                ],
                "timestamp": 1607879735685,
              },
            },
          },
          "entry": Object {
            "isDirectory": true,
            "owners": Array [
              "ROLE_USER",
            ],
            "readers": Array [
              "ROLE_USER",
            ],
          },
        },
        "thumbnails": Object {
          "children": Object {},
          "entry": Object {
            "isDirectory": true,
            "owners": Array [
              "ROLE_USER",
            ],
            "readers": Array [
              "ROLE_USER",
            ],
          },
        },
      },
      "entry": Object {
        "isDirectory": true,
        "owners": Array [
          "ROLE_CS_ROOT",
        ],
        "readers": Array [
          "ROLE_CS_ROOT",
        ],
      },
    },
    "filters": Object {
      "children": Object {
        "content": Object {
          "children": Object {},
          "entry": Object {
            "isDirectory": true,
            "owners": Array [
              "ROLE_USER",
            ],
            "readers": Array [
              "ROLE_USER",
            ],
          },
        },
        "structure": Object {
          "children": Object {},
          "entry": Object {
            "isDirectory": true,
            "owners": Array [
              "ROLE_USER",
            ],
            "readers": Array [
              "ROLE_USER",
            ],
          },
        },
      },
      "entry": Object {
        "isDirectory": true,
        "owners": Array [
          "ROLE_CS_ROOT",
        ],
        "readers": Array [
          "ROLE_CS_ROOT",
        ],
      },
    },
    "organization_permissions": Object {
      "entry": Object {
        "content": "{}",
        "isDirectory": false,
        "owners": Array [
          "ROLE_CS_ROOT",
        ],
        "readers": Array [
          "ROLE_CS_ROOT",
        ],
      },
    },
    "organization_settings": Object {
      "entry": Object {
        "content": "{}",
        "owners": Array [
          "ROLE_CS_ROOT",
        ],
        "readers": Array [
          "ROLE_USER",
        ],
      },
    },
    "user_roles": Object {
      "entry": Object {
        "isDirectory": true,
        "owners": Array [
          "ROLE_CS_ROOT",
        ],
        "readers": Array [
          "ROLE_USER",
        ],
      },
    },
    "users": Object {
      "children": Object {
        "admin": Object {
          "children": Object {
            "activity": Object {
              "entry": Object {
                "content": "{\\"userFilters\\":[\\"[Geography].[City].[ALL].[AllMember].[Berlin]\\"]}",
                "owners": Array [
                  "admin",
                ],
                "readers": Array [
                  "admin",
                ],
              },
            },
            "settings": Object {
              "entry": Object {
                "content": "{\\"theme\\":\\"dark-activeviam\\",\\"search.maxResults\\":10,\\"userFilters.areEnabled\\":true}",
                "owners": Array [
                  "admin",
                ],
                "readers": Array [
                  "admin",
                ],
              },
            },
          },
          "entry": Object {
            "canRead": true,
            "canWrite": true,
            "isDirectory": true,
            "lastEditor": "admin",
            "owners": Array [
              "admin",
            ],
            "readers": Array [
              "admin",
            ],
            "timestamp": 1607879735685,
          },
        },
        "user1": Object {
          "children": Object {
            "activity": Object {
              "entry": Object {
                "content": "{}",
                "owners": Array [
                  "user1",
                ],
                "readers": Array [
                  "user1",
                ],
              },
            },
            "settings": Object {
              "entry": Object {
                "content": "{}",
                "owners": Array [
                  "user1",
                ],
                "readers": Array [
                  "user1",
                ],
              },
            },
          },
          "entry": Object {
            "canRead": true,
            "canWrite": true,
            "isDirectory": true,
            "lastEditor": "admin",
            "owners": Array [
              "user1",
            ],
            "readers": Array [
              "user1",
            ],
            "timestamp": 1607879735685,
          },
        },
      },
      "entry": Object {
        "isDirectory": true,
        "owners": Array [
          "ROLE_USER",
        ],
        "readers": Array [
          "ROLE_USER",
        ],
      },
    },
    "widgets": Object {
      "children": Object {
        "content": Object {
          "children": Object {},
          "entry": Object {
            "isDirectory": true,
            "owners": Array [
              "ROLE_USER",
            ],
            "readers": Array [
              "ROLE_USER",
            ],
          },
        },
        "structure": Object {
          "children": Object {},
          "entry": Object {
            "isDirectory": true,
            "owners": Array [
              "ROLE_USER",
            ],
            "readers": Array [
              "ROLE_USER",
            ],
          },
        },
      },
      "entry": Object {
        "isDirectory": true,
        "owners": Array [
          "ROLE_CS_ROOT",
        ],
        "readers": Array [
          "ROLE_CS_ROOT",
        ],
      },
    },
  },
  "entry": Object {
    "isDirectory": true,
    "owners": Array [
      "ROLE_CS_ROOT",
    ],
    "readers": Array [
      "ROLE_CS_ROOT",
    ],
  },
}
`;

exports[`migratePivotFolder returns a valid ActiveUI5 /ui folder on an input that does not contain an entry for calculated measures 1`] = `
Object {
  "children": Object {
    "calculated_measures": Object {
      "children": Object {
        "content": Object {
          "children": Object {
            "003": Object {
              "entry": Object {
                "content": "{\\"expression\\":\\"IIf(IsEmpty(([Booking].[Desk].CurrentMember.Parent, [Measures].[pnl.SUM])), null, [Measures].[pnl.SUM] / ([Booking].[Desk].CurrentMember.Parent, [Measures].[pnl.SUM]))\\",\\"properties\\":[\\"FORMAT_STRING=\\\\\\"#,###.##%\\\\\\"\\"]}",
                "isDirectory": false,
              },
            },
            "004": Object {
              "entry": Object {
                "content": "{\\"expression\\":\\"IIf(IsEmpty(([Currency].[Currency].[ALL].[AllMember], [Measures].[pnl.FOREX])), null, [Measures].[pnl.FOREX] / ([Currency].[Currency].[ALL].[AllMember], [Measures].[pnl.FOREX]))\\",\\"properties\\":[\\"FORMAT_STRING=\\\\\\"#,###.##%\\\\\\"\\"]}",
                "isDirectory": false,
              },
            },
            "005": Object {
              "entry": Object {
                "content": "{\\"expression\\":\\"Count(Descendants([CounterParty].[CounterParty].CurrentMember, [CounterParty].[CounterParty].[CounterPartyGroup]), EXCLUDEEMPTY)\\",\\"properties\\":[\\"FORMAT_STRING=\\\\\\"#,###.##\\\\\\"\\"]}",
                "isDirectory": false,
              },
            },
          },
          "entry": Object {
            "isDirectory": true,
            "owners": Array [
              "ROLE_USER",
            ],
            "readers": Array [
              "ROLE_USER",
            ],
          },
        },
        "structure": Object {
          "children": Object {
            "003": Object {
              "children": Object {
                "003_metadata": Object {
                  "entry": Object {
                    "content": "{\\"name\\":\\"[Measures].[third calculated measure]\\"}",
                    "isDirectory": false,
                  },
                },
              },
              "entry": Object {
                "isDirectory": true,
              },
            },
            "004": Object {
              "children": Object {
                "004_metadata": Object {
                  "entry": Object {
                    "content": "{\\"name\\":\\"[Measures].[second calculated measure]\\"}",
                    "isDirectory": false,
                  },
                },
              },
              "entry": Object {
                "isDirectory": true,
              },
            },
            "005": Object {
              "children": Object {
                "005_metadata": Object {
                  "entry": Object {
                    "content": "{\\"name\\":\\"[Measures].[first calculated measure]\\"}",
                    "isDirectory": false,
                  },
                },
              },
              "entry": Object {
                "isDirectory": true,
              },
            },
          },
          "entry": Object {
            "isDirectory": true,
            "owners": Array [
              "ROLE_USER",
            ],
            "readers": Array [
              "ROLE_USER",
            ],
          },
        },
      },
      "entry": Object {
        "isDirectory": true,
        "owners": Array [
          "ROLE_USER",
        ],
        "readers": Array [
          "ROLE_USER",
        ],
      },
    },
    "dashboards": Object {
      "children": Object {
        "content": Object {
          "children": Object {
            "bef": Object {
              "entry": Object {
                "canRead": true,
                "canWrite": true,
                "content": "{\\"pages\\":{\\"p-0\\":{\\"layout\\":{\\"children\\":[{\\"leafKey\\":\\"0\\",\\"size\\":1}],\\"direction\\":\\"row\\"},\\"name\\":\\"Page 1\\",\\"filters\\":[],\\"content\\":{\\"0\\":{\\"query\\":{\\"mdx\\":\\"SELECT FROM [EquityDerivativesCube] CELL PROPERTIES VALUE, FORMATTED_VALUE, BACK_COLOR, FORE_COLOR, FONT_FLAGS\\",\\"updateMode\\":\\"once\\"},\\"filters\\":[],\\"queryContext\\":[],\\"mapping\\":{\\"rows\\":[],\\"columns\\":[\\"ALL_MEASURES\\"],\\"measures\\":[]},\\"name\\":\\"Pivot table with calculated measures\\",\\"serverKey\\":\\"my-server\\",\\"widgetKey\\":\\"tree-table\\",\\"columnWidths\\":{}}},\\"queryContext\\":[]}},\\"pagesOrder\\":[\\"p-0\\"],\\"filters\\":[],\\"queryContext\\":[]}",
                "isDirectory": false,
                "lastEditor": "admin",
                "owners": Array [
                  "admin",
                ],
                "readers": Array [
                  "admin",
                ],
                "timestamp": 1658155953186,
              },
            },
          },
          "entry": Object {
            "isDirectory": true,
            "owners": Array [
              "ROLE_USER",
            ],
            "readers": Array [
              "ROLE_USER",
            ],
          },
        },
        "structure": Object {
          "children": Object {
            "bef": Object {
              "children": Object {
                "bef_metadata": Object {
                  "entry": Object {
                    "canRead": true,
                    "canWrite": true,
                    "content": "{\\"name\\":\\"Calculated measures dashboard.\\"}",
                    "isDirectory": true,
                    "lastEditor": "admin",
                    "owners": Array [
                      "admin",
                    ],
                    "readers": Array [
                      "admin",
                    ],
                    "timestamp": 1658155953644,
                  },
                },
              },
              "entry": Object {
                "canRead": true,
                "canWrite": true,
                "isDirectory": true,
                "lastEditor": "admin",
                "owners": Array [
                  "admin",
                ],
                "readers": Array [
                  "admin",
                ],
                "timestamp": 1658155953644,
              },
            },
          },
          "entry": Object {
            "isDirectory": true,
            "owners": Array [
              "ROLE_USER",
            ],
            "readers": Array [
              "ROLE_USER",
            ],
          },
        },
        "thumbnails": Object {
          "children": Object {},
          "entry": Object {
            "isDirectory": true,
            "owners": Array [
              "ROLE_USER",
            ],
            "readers": Array [
              "ROLE_USER",
            ],
          },
        },
      },
      "entry": Object {
        "isDirectory": true,
        "owners": Array [
          "ROLE_CS_ROOT",
        ],
        "readers": Array [
          "ROLE_CS_ROOT",
        ],
      },
    },
    "filters": Object {
      "children": Object {
        "content": Object {
          "children": Object {},
          "entry": Object {
            "isDirectory": true,
            "owners": Array [
              "ROLE_USER",
            ],
            "readers": Array [
              "ROLE_USER",
            ],
          },
        },
        "structure": Object {
          "children": Object {},
          "entry": Object {
            "isDirectory": true,
            "owners": Array [
              "ROLE_USER",
            ],
            "readers": Array [
              "ROLE_USER",
            ],
          },
        },
      },
      "entry": Object {
        "isDirectory": true,
        "owners": Array [
          "ROLE_CS_ROOT",
        ],
        "readers": Array [
          "ROLE_CS_ROOT",
        ],
      },
    },
    "organization_permissions": Object {
      "entry": Object {
        "content": "{}",
        "isDirectory": false,
        "owners": Array [
          "ROLE_CS_ROOT",
        ],
        "readers": Array [
          "ROLE_CS_ROOT",
        ],
      },
    },
    "organization_settings": Object {
      "entry": Object {
        "content": "{}",
        "owners": Array [
          "ROLE_CS_ROOT",
        ],
        "readers": Array [
          "ROLE_USER",
        ],
      },
    },
    "user_roles": Object {
      "entry": Object {
        "isDirectory": true,
        "owners": Array [
          "ROLE_CS_ROOT",
        ],
        "readers": Array [
          "ROLE_USER",
        ],
      },
    },
    "users": Object {
      "children": Object {
        "admin": Object {
          "children": Object {
            "activity": Object {
              "entry": Object {
                "content": "{}",
                "owners": Array [
                  "admin",
                ],
                "readers": Array [
                  "admin",
                ],
              },
            },
            "settings": Object {
              "entry": Object {
                "content": "{}",
                "owners": Array [
                  "admin",
                ],
                "readers": Array [
                  "admin",
                ],
              },
            },
          },
          "entry": Object {
            "canRead": true,
            "canWrite": true,
            "isDirectory": true,
            "lastEditor": "admin",
            "owners": Array [
              "admin",
            ],
            "readers": Array [
              "admin",
            ],
            "timestamp": 1658155953761,
          },
        },
      },
      "entry": Object {
        "isDirectory": true,
        "owners": Array [
          "ROLE_USER",
        ],
        "readers": Array [
          "ROLE_USER",
        ],
      },
    },
    "widgets": Object {
      "children": Object {
        "content": Object {
          "children": Object {},
          "entry": Object {
            "isDirectory": true,
            "owners": Array [
              "ROLE_USER",
            ],
            "readers": Array [
              "ROLE_USER",
            ],
          },
        },
        "structure": Object {
          "children": Object {},
          "entry": Object {
            "isDirectory": true,
            "owners": Array [
              "ROLE_USER",
            ],
            "readers": Array [
              "ROLE_USER",
            ],
          },
        },
      },
      "entry": Object {
        "isDirectory": true,
        "owners": Array [
          "ROLE_CS_ROOT",
        ],
        "readers": Array [
          "ROLE_CS_ROOT",
        ],
      },
    },
  },
  "entry": Object {
    "isDirectory": true,
    "owners": Array [
      "ROLE_CS_ROOT",
    ],
    "readers": Array [
      "ROLE_CS_ROOT",
    ],
  },
}
`;
